### PR TITLE
solid was harder

### PR DESCRIPTION
**Pull request by IhorRome**

### CODE REVIEW

File: src/main/java/mate/jdbc/Main.java [hunk 1]
```diff
@@ -1,7 +1,32 @@
...
+        manufacturerToCreate.setName("Dell");
+        manufacturerToCreate.setCountry("US");
+        Manufacturer createdManufacturer = dao.create(manufacturerToCreate);
+
+        Manufacturer manufacturerGotById = dao.get(3L).orElseThrow();
+
+        boolean isDeleted = dao.delete(5L);
+
+        Manufacturer manufacturerToUpdate = new Manufacturer();
+        manufacturerToUpdate.setId(2L);
```
Comment (kupora3) [inline src/main/java/mate/jdbc/Main.java:23]: Не понятно, что такое 5L 3L 2L, наверное лучше это вынести в константы или вызывать ИД из самой мануфактуры методом getID после отработки create ид у объекта присвоится

**Review by kupora3 — CHANGES_REQUESTED** [review]

**Review by kupora3 — APPROVED** [review]

File: src/main/java/mate/jdbc/Main.java [hunk 2]
```diff
@@ -1,7 +1,37 @@
...
+        dell.setName("Dell");
+        dell.setCountry("US");
+        Manufacturer honda = new Manufacturer();
+        dell.setName("Honda");
+        dell.setCountry("Japan");
+
+        Manufacturer uploadedDell = dao.create(dell);
+        Manufacturer uploadedHonda = dao.create(honda);
+
+        Manufacturer manufacturerGotById = dao.get(uploadedDell.getId()).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/Main.java:23]: let's print it out?

File: src/main/java/mate/jdbc/Main.java [hunk 3]
```diff
@@ -1,7 +1,37 @@
...
+
+        boolean isDeleted = dao.delete(uploadedDell.getId());
+
+        Manufacturer apple = new Manufacturer();
+        apple.setId(uploadedHonda.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/Main.java:28]: this is not how we are supposed to do this

we cannot update the object that is not yet in our database
you should change fields of the object that was already saved with create() method
and then call update()

and you should not update id

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement insertManufacturerStatement = connection
+                        .prepareStatement(insertManufacturerRequest,
+                                Statement.RETURN_GENERATED_KEYS)) {
+            insertManufacturerStatement.setString(1, manufacturer.getName());
+            insertManufacturerStatement.setString(2, manufacturer.getCountry());
+            insertManufacturerStatement.executeUpdate();
+            ResultSet generatedKeys = insertManufacturerStatement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+                Long id = generatedKeys.getLong(1);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:30]: please see checklist

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            deleteManufacturerStatement.setLong(1, id);
+            return deleteManufacturerStatement.executeUpdate() > 0;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't delete manufacturer by Id = " + id, e);
+        }
+    }
+
+    private Manufacturer getManufacturer(ResultSet resultSet) throws SQLException {
+        Manufacturer manufacturer = new Manufacturer();
+        manufacturer.setId(resultSet.getObject(1, Long.class));
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:116]:
```suggestion
        manufacturer.setId(resultSet.getObject("id", Long.class));
```

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't insert manufacturer" + manufacturer, e);
+        }
+        return manufacturer;
+    }
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        String selectManufacturerRequest =
+                "SELECT id, name, country FROM manufacturers WHERE id = ? AND is_deleted = false";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:42]:
```suggestion
                "SELECT * FROM manufacturers WHERE id = ? AND is_deleted = false";
```

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get manufacturer by Id = " + id, e);
+        }
+
+    }
+
+    @Override
+    public List<Manufacturer> getAll() {
+        String selectAllManufacturersRequest =
+                "SELECT id, name, country FROM manufacturers WHERE is_deleted = false;";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:62]: same here

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                Statement selectAllStatement = connection.createStatement()) {
+            ResultSet resultSet = selectAllStatement.executeQuery(selectAllManufacturersRequest);
+            List<Manufacturer> tableData = new ArrayList<>();
+            Manufacturer manufacturer = null;
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:67]: do we need this variable here?

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            while (resultSet.next()) {
+                manufacturer = getManufacturer(resultSet);
+                tableData.add(manufacturer);
+            }
+            return tableData;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get all data from DB", e);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:74]:
```suggestion
            throw new DataProcessingException("Can't get all manufacturers from DB", e);
```

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        String updateManufacturerRequest =
+                "UPDATE manufacturers SET name = ?, country = ? "
+                        + "WHERE id = ? AND is_deleted = false;";
+        Manufacturer updatedManufacturer = new Manufacturer();
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:84]: why do you need this variable? you already receive the updated manufacturer in the method

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement updateManufacturerStatement = connection
+                        .prepareStatement(updateManufacturerRequest)) {
+            updateManufacturerStatement.setString(1, manufacturer.getName());
+            updateManufacturerStatement.setString(2, manufacturer.getCountry());
+            updateManufacturerStatement.setLong(3, manufacturer.getId());
+            updateManufacturerStatement.executeUpdate();
+            updatedManufacturer = get(manufacturer.getId()).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:92]:
```suggestion
            return manufacturer;
```

File: src/main/java/mate/jdbc/util/ConnectionUtil.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
+package mate.jdbc.util;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.Properties;
+import mate.jdbc.exception.DataProcessingException;
+
+public class ConnectionUtil {
+
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/util/ConnectionUtil.java:10]: redundant empty line

File: src/main/java/mate/jdbc/util/ConnectionUtil.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    private static final String DRIVER_CLASS_NAME = "com.mysql.cj.jdbc.Driver";
+    private static final String USER_NAME = "business_user";
+    private static final String USER_PASSWORD = "87654321";
+    private static final String URL_TO_DB = "jdbc:mysql://localhost:3306/business_db";
+
+    static {
+        try {
+            Class.forName(DRIVER_CLASS_NAME);
+        } catch (ClassNotFoundException e) {
+            throw new DataProcessingException("Can't load JDBC driver ", e);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/util/ConnectionUtil.java:20]: it is better to throw DataProcessingException only in dao
here we can use RuntimeException

File: src/main/java/mate/jdbc/util/ConnectionUtil.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+    public static Connection getConnection() {
+        try {
+            Properties properties = new Properties();
+            properties.put("user", USER_NAME);
+            properties.put("password", USER_PASSWORD);
+            return DriverManager
+                    .getConnection(URL_TO_DB, properties);
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get connection with DB ", e);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/jdbc/util/ConnectionUtil.java:32]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
let's create the `init_db.sql` as well

File: src/main/resources/init_file.sql [hunk 15]
```diff
@@ -0,0 +1,8 @@
+CREATE DATABASE `business_db` /*!40100 DEFAULT CHARACTER SET utf8 */ /*!80016 DEFAULT ENCRYPTION='N' */;
+CREATE TABLE `manufacturers` (
+                                 `id` int NOT NULL AUTO_INCREMENT,
```
Comment (KseniiaMakarova) [inline src/main/resources/init_file.sql:3]: use `bigint` - not `int`

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
please also check out some hidden conversations
some of the comments were not fixed

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**